### DISCUSSION TITLE
rtx doesn't update GOPATH in GoLand's terminal

### DISCUSSION THREAD

**Discussion by Crocmagnon** [jdx/mise#815] [General]
I'm using rtx with fish, it works fine in my usual terminal emulator but fails to update GOPATH and GOROOT in GoLand's integrated terminal emulator.
However, if I launch GoLand from the shell script provided with the JetBrains Toolbox, it works fine.

I have setup `rtx activate` in my fish, bash and zsh config:

```console
$ cat ~/.config/fish/config.fish | grep "rtx activate"
rtx activate fish | source

$ cat ~/.bashrc | grep "rtx activate"
eval "$(~/bin/rtx activate bash)"

$ cat ~/.zshrc | grep "rtx activate"
eval "$(~/bin/rtx activate zsh)"
```

Did I miss something in my config? Is it a known issue? I read the README about IDE integration but it looks like it applies to running binaries from the IDE, not from terminal emulators in the IDE.

<details>
<summary>In iTerm2</summary>

```console
$ rtx hook-env
set -e GOPATH
set -e GOROOT
set -gx PATH '/Users/gaugendr/go/bin:/Library/Frameworks/Python.framework/Versions/3.10/bin:/Users/gaugendr/Library/Python/3.10/bin:/Users/gaugendr/Library/Application Support/JetBrains/Toolbox/scripts:/Users/gaugendr/.krew/bin:/Users/gaugendr/.local/bin:/Users/gaugendr/.fzf/bin:/usr/local/bin:/System/Cryptexes/App/usr/bin:/usr/bin:/bin:/usr/sbin:/sbin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/local/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/appleinternal/bin:/usr/local/MacGPG2/bin'
set -gx GOROOT /Users/gaugendr/.local/share/rtx/installs/go/1.20.7/go
set -gx GOPATH /Users/gaugendr/.local/share/rtx/installs/go/1.20.7/packages
set -gx PATH '/Users/gaugendr/.local/share/rtx/installs/go/1.20.7/go/bin:/Users/gaugendr/.local/share/rtx/installs/go/1.20.7/packages/bin:/Users/gaugendr/go/bin:/Library/Frameworks/Python.framework/Versions/3.10/bin:/Users/gaugendr/Library/Python/3.10/bin:/Users/gaugendr/Library/Application Support/JetBrains/Toolbox/scripts:/Users/gaugendr/.krew/bin:/Users/gaugendr/.local/bin:/Users/gaugendr/.fzf/bin:/usr/local/bin:/System/Cryptexes/App/usr/bin:/usr/bin:/bin:/usr/sbin:/sbin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/local/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/appleinternal/bin:/usr/local/MacGPG2/bin'
set -gx __RTX_DIFF eAGdzbENwjAQheGNfEABEqKACjojFAY4ktMl4uSLfEZQItiAFYLYydsQChZw/77/PQeV5j4Euj4+W3/wvspzOBpFA8YLU2giONEaBazFSBDTDbpgCUXGicLUzSZuAawj32+qXV6V8B7rMzLZu8fUvvKypMEKpy7kdYn9//8KX2epZMM
set -gx __RTX_WATCH eAFtzD8KwjAUgHEnxdUTuLk1WCniQdyEkCYvfyC+SF4tzpk8g7ubQzyGi3tBPICnEHESu318wy+dtfNAqZuxNUEkZsTeAKrIChlQO8Nic2DfLJqw9SkTSOLkUAKHXZD2rp6D4xUFht89Om2G3fSP9UEKz8iKCB+715s8erzx65YvgC1vReRWkM1lXatFBaVarvRcy+oNIelU5Q

$ rtx doctor
rtx version:
  2023.8.1 macos-arm64 (f64635c 2023-08-17)

build:
  Target: aarch64-apple-darwin
  Features: DEFAULT, NATIVE_TLS, OPENSSL, SELF_UPDATE
  Built: Thu, 17 Aug 2023 01:42:25 +0000
  Rust Version: rustc 1.71.1 (eb26296b5 2023-08-03)
  Profile: release

shell:
  /usr/local/bin/fish
  fish, version 3.6.1

rtx data directory:
  /Users/gaugendr/.local/share/rtx

rtx environment variables:
  RTX_SHELL=fish

settings:
  {"always_keep_download": "false", "always_keep_install": "false", "asdf_compat": "false", "disable_default_shorthands": "false", "disable_tools": "[]", "experimental": "false", "jobs": "4", "legacy_version_file": "true", "legacy_version_file_disable_tools": "[]", "log_level": "INFO", "missing_runtime_behavior": "warn", "plugin_autoupdate_last_check_duration": "10080", "raw": "false", "trusted_config_paths": "[]", "verbose": "false", "yes": "false"}

config files:
  /Users/gaugendr/.config/rtx/config.toml

plugins:
  go              (core)
  golangci-lint   https://github.com/hypnoglow/asdf-golangci-lint.git#a43c37c
  java            (core)
  node            https://github.com/rtx-plugins/rtx-nodejs.git#da3fc45
  python          (core)
  ruby            (core)

toolset:
  go@prefix:1.20

No problems found
```
</details>

<details>
<summary>In Goland (opened as a normal app)</summary>

```console
$ rtx hook-env
set -gx PATH '/Users/gaugendr/.local/share/rtx/installs/go/1.20.7/go/bin:/Users/gaugendr/go/bin:/Users/gaugendr/go/bin:/Library/Frameworks/Python.framework/Versions/3.10/bin:/Users/gaugendr/Library/Python/3.10/bin:/Users/gaugendr/Library/Application Support/JetBrains/Toolbox/scripts:/Users/gaugendr/.krew/bin:/Users/gaugendr/.local/bin:/Users/gaugendr/.fzf/bin:/usr/local/bin:/System/Cryptexes/App/usr/bin:/usr/bin:/bin:/usr/sbin:/sbin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/local/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/appleinternal/bin:/usr/local/MacGPG2/bin'
set -gx PATH '/Users/gaugendr/.local/share/rtx/installs/go/1.20.7/go/bin:/Users/gaugendr/.local/share/rtx/installs/go/1.20.7/packages/bin:/Users/gaugendr/.local/share/rtx/installs/go/1.20.7/go/bin:/Users/gaugendr/go/bin:/Users/gaugendr/go/bin:/Library/Frameworks/Python.framework/Versions/3.10/bin:/Users/gaugendr/Library/Python/3.10/bin:/Users/gaugendr/Library/Application Support/JetBrains/Toolbox/scripts:/Users/gaugendr/.krew/bin:/Users/gaugendr/.local/bin:/Users/gaugendr/.fzf/bin:/usr/local/bin:/System/Cryptexes/App/usr/bin:/usr/bin:/bin:/usr/sbin:/sbin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/local/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/appleinternal/bin:/usr/local/MacGPG2/bin'
set -gx __RTX_DIFF eAGdzMENQEAQRuGO9sdF4qQJBQwmS0x2ZWeFo0QHWqCp7QYtuL/3HZeXfr8cr/s9UxzOVKFRDgpLi2XXBxjxHQl0oMAIccPoNJLIm3jkpshMCevRji7Vf96Zuoks6yc8w1M1cw
set -gx __RTX_WATCH eAFtzDEKwjAUgGEnBWcP4ObWYKWIB3ETQvry0gTii+TV4pypZ3Dv5lAv4t7Fe7iJOIndfv7hS51xHjkNK7FnjCwqda6QdBQZBDKuErG+iG9mdTj61DMCS3YEKPEUwD70c9LeSVH43bPrYTos/1gfQHnBVkX82KPeQo9481fX3pAa2agorWLb52WpNwXmerszawPFGwHAVDM

$  rtx doctor
rtx version:
  2023.8.1 macos-arm64 (f64635c 2023-08-17)

build:
  Target: aarch64-apple-darwin
  Features: DEFAULT, NATIVE_TLS, OPENSSL, SELF_UPDATE
  Built: Thu, 17 Aug 2023 01:42:25 +0000
  Rust Version: rustc 1.71.1 (eb26296b5 2023-08-03)
  Profile: release

shell:
  /usr/local/bin/fish
  fish, version 3.6.1

rtx data directory:
  /Users/gaugendr/.local/share/rtx

rtx environment variables:
  RTX_SHELL=fish

settings:
  {"always_keep_download": "false", "always_keep_install": "false", "asdf_compat": "false", "disable_default_shorthands": "false", "disable_tools": "[]", "experimental": "false", "jobs": "4", "legacy_version_file": "true", "legacy_version_file_disable_tools": "[]", "log_level": "INFO", "missing_runtime_behavior": "warn", "plugin_autoupdate_last_check_duration": "10080", "raw": "false", "trusted_config_paths": "[]", "verbose": "false", "yes": "false"}

config files:
  /Users/gaugendr/.config/rtx/config.toml

plugins:
  go              (core)
  golangci-lint   https://github.com/hypnoglow/asdf-golangci-lint.git#a43c37c
  java            (core)
  node            https://github.com/rtx-plugins/rtx-nodejs.git#da3fc45
  python          (core)
  ruby            (core)

toolset:
  go@prefix:1.20

No problems found
```
</details>

<details>
<summary>In Goland (opened with the shell alias)</summary>

```console
$ rtx hook-env
set -e GOPATH
set -e GOROOT
set -gx PATH '/Users/gaugendr/go/bin:/Users/gaugendr/go/bin:/Library/Frameworks/Python.framework/Versions/3.10/bin:/Users/gaugendr/Library/Python/3.10/bin:/Users/gaugendr/Library/Application Support/JetBrains/Toolbox/scripts:/Users/gaugendr/.krew/bin:/Users/gaugendr/.local/bin:/Users/gaugendr/.fzf/bin:/usr/local/bin:/System/Cryptexes/App/usr/bin:/usr/bin:/bin:/usr/sbin:/sbin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/local/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/appleinternal/bin:/usr/local/MacGPG2/bin:/Users/gaugendr/.local/share/rtx/installs/go/1.20.7/go/bin'
set -gx GOPATH /Users/gaugendr/.local/share/rtx/installs/go/1.20.7/packages
set -gx GOROOT /Users/gaugendr/.local/share/rtx/installs/go/1.20.7/go
set -gx PATH '/Users/gaugendr/.local/share/rtx/installs/go/1.20.7/go/bin:/Users/gaugendr/.local/share/rtx/installs/go/1.20.7/packages/bin:/Users/gaugendr/go/bin:/Users/gaugendr/go/bin:/Library/Frameworks/Python.framework/Versions/3.10/bin:/Users/gaugendr/Library/Python/3.10/bin:/Users/gaugendr/Library/Application Support/JetBrains/Toolbox/scripts:/Users/gaugendr/.krew/bin:/Users/gaugendr/.local/bin:/Users/gaugendr/.fzf/bin:/usr/local/bin:/System/Cryptexes/App/usr/bin:/usr/bin:/bin:/usr/sbin:/sbin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/local/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/appleinternal/bin:/usr/local/MacGPG2/bin:/Users/gaugendr/.local/share/rtx/installs/go/1.20.7/go/bin'
set -gx __RTX_DIFF eAGdzrENwjAQheGN/IACJEQBFXRGKAxwJCcn4mRHPiMoEWzACkHs5G0IBQu4f/+n9xyCNPfB8/Xx2dr9ptrlFY7KUeHo4tg3EUZCTQJtKTJiuqHzmkhknARMzWxiFuipPpNjHZGDtVWelyAuvHtK7Ssvy2qcOp/XJe3//k/4AnRvZMM
set -gx __RTX_WATCH eAFtzD8KwjAUgHFB0At4ADe3BqtFPIibENLk5Q/EF8mrxTlTR2d3N4d6Efcuzh5DxEna7eMbfummnQdK3YLtCCIxI04GUEWWyYDaGRarM/tlVoWDTy2BJE4OJXA4Bmmf6jVqHigw/O/pdT/p5j3WByk8IysifO1Bb/Ye8MbN+nIHrHktIreCbJuXpVoVkKvNVi+1LD4A8VPW

$ rtx doctor
rtx version:
  2023.8.1 macos-arm64 (f64635c 2023-08-17)

build:
  Target: aarch64-apple-darwin
  Features: DEFAULT, NATIVE_TLS, OPENSSL, SELF_UPDATE
  Built: Thu, 17 Aug 2023 01:42:25 +0000
  Rust Version: rustc 1.71.1 (eb26296b5 2023-08-03)
  Profile: release

shell:
  /usr/local/bin/fish
  fish, version 3.6.1

rtx data directory:
  /Users/gaugendr/.local/share/rtx

rtx environment variables:
  RTX_SHELL=fish

settings:
  {"always_keep_download": "false", "always_keep_install": "false", "asdf_compat": "false", "disable_default_shorthands": "false", "disable_tools": "[]", "experimental": "false", "jobs": "4", "legacy_version_file": "true", "legacy_version_file_disable_tools": "[]", "log_level": "INFO", "missing_runtime_behavior": "warn", "plugin_autoupdate_last_check_duration": "10080", "raw": "false", "trusted_config_paths": "[]", "verbose": "false", "yes": "false"}

config files:
  /Users/gaugendr/.config/rtx/config.toml

plugins:
  go              (core)
  golangci-lint   https://github.com/hypnoglow/asdf-golangci-lint.git#a43c37c
  java            (core)
  node            https://github.com/rtx-plugins/rtx-nodejs.git#da3fc45
  python          (core)
  ruby            (core)

toolset:
  go@prefix:1.20

No problems found
```
</details>

Adding "IntellIJ IDEA" for search purposes.

**Comment by jdx** [comment]
I'm pretty sure what is happening is Goland sets GOPATH itself. If rtx sets it before, then it feels at liberty to update it but if another tool (or zshrc) sets GOPATH then it does not screw with it. I inherited this behavior from the asdf plugin, but I'm not sure if it's wise or not. It seems there are use-cases where people _want_ to have GOPATH overridden (or even set at all) and use-cases where they don't.

What we could do is update this logic a bit to force GOPATH setting if RTX_GO_SET_GOROOT is true: https://github.com/jdxcode/rtx/blob/06f2bef728fc739e8668ce6b9c14134b697e8561/src/plugins/core/go.rs#L182-L198 I'd accept a PR for that.

**Reply by jdx** [reply]
See https://github.com/jdxcode/rtx/issues/754

cc @scop

**Reply by scop** [reply]
As long as there's a way to tell rtx not touch anything Go related but PATH, works for me :) The suggested additional semantics (I gather true => always set, no matter what) make sense to me, too.

**Reply by jdx** [reply]
I also fear that the defaults are bad, but not sure if I should change it

**Reply by Crocmagnon** [reply]
I'm sorry, I'm unsure what to do. Is there something I can already configure to fix my issue? Or does this require development? If so, maybe we should convert this to an issue? I'm not familiar enough with rust to provide an implementation myself.

**Reply by brigitops** [reply]
I think GOPATH and GOROOT are different cases, and therefore should have different default behaviours as to whether to override them:

GOPATH is something that you often just leave at $HOME/go
GOROOT is expected to point to the go tool that you want to use

I got a surprise with GOROOT being set by GoLand in the terminal and not by mise. The following failed because the go tool version doesn't match the GOROOT:

```shell
mise install
mise exec -- go install gotest.tools/gotestsum@v1.11.0
```

In the principle of least astonishment, I think that this should ensure that GOROOT corresponds to the tool chosen by mise. On balance, it's safer to override than leaving the environment variable alone, because we lose the reproducibility we're expecting from mise exec, and we don't even realise it! Particularly that could be nasty in continuous integration environments.

If you look at docs like https://go.dev/doc/manage-install for managing multiple versions of go, the GOROOT belongs to the go version.

My suggestion, then, would be to have options to override both variables, and then defaults as follows:

override GOPATH by default: no
override GOROOT by default: yes

**Reply by jdx** [reply]
it looks like go automatically sets it:

```
usage  pkl  2 ❯ mise x go@latest -- go env GOROOT
/Users/jdx/.mise/installs/go/1.20.13/go
usage  pkl  2 ❯ mise x -- go env GOROOT
/opt/homebrew/Cellar/go/1.21.6/libexec
```

I think leaving this as an overridable option still makes the most sense

**Reply by brigitops** [reply]
I think this is because you mustn't already have it set. Whereas, this discussion is around an environment where it's already been set (by GoLand). If it's already set, then just like mise go will leave the GOROOT alone

This leads to the astonishing result that mise exec doesn't actually switch the tool properly:

```
$ go env GOROOT
/home/.../sdk/go1.21.6

$ mise exec -- go env GOROOT
/home/.../sdk/go1.21.6

$mise exec go@1.22 -- go env GOROOT
/home/.../sdk/go1.21.6
```

Having at least the option to solve this would be awesome, the question of the default (which I still think is wrong) can be dealt with later if we have the options in the meantime. Right now there's a big gotcha. My priority would be to put something in .mise.toml to make mise always set correctly (or even unset) the GOROOT so people can use mise easily from GoLand terminal

**Reply by jdx** [reply]
huh I thought we had made a decision to default both of these to not be set by default, which I still think is what we should do even if it makes goland a bit tougher to configure. That seems to not be the case if the docs are correct though.

But I agree with you that there should be an option to override, MISE_GO_SET_GOROOT=1 should not simply set it if it does not exist but also overwrite any existing value.

**Reply by jdx** [reply]
The fact the default sets GOROOT is making me think we shouldn't change the default, let me chat with the advisory board about this one.

In other words we should keep setting the env var but tweak overwriting behavior 

I thought current behavior was the opposite

**Reply by jdx** [reply]
Looking at some of the history around this it seems there are reasons to set GOROOT explicitly even though `go` defaults it: https://github.com/asdf-community/asdf-golang/issues/28

I'm not sure I follow why people want to set GOBIN and GOPATH to the asdf directory though

**Reply by brigitops** [reply]
Ok, right, I thought the behaviour was something else, too!

Does look like this is really a matter of clearing up documented/intended/implemented behaviour.

I was thinking of mise as always inheriting the host environment, and that it would never touch GOROOT if it were already set in the host environment. I.e. that regardless of what I do with MISE_GO_SET_GOROOT it will stick with the value set by GoLand terminal, meaning I have to make sure to unset it myself.

Whereas, MISE_GO_SET_GOROOT=true solves my problem:

```.mise.toml
[tools]
go = "1.22.0"

[env]
MISE_GO_SET_GOROOT="true"
```

Now gives the result I expected in the GoLand terminal:

```
$mise exec go@1.22 -- go env GOROOT
/home/.../.local/share/mise/installs/go/1.22.0/go
```

The documentation claims that the default is true, but it doesn't seem to be then, I get a different behaviour by setting it explicitly to true?

Maybe, even if default doesn't change, we can add a section to the documentation for GoLand users to suggest setting MISE_GO_SET_GOROOT=true in the .mise.toml?

And for my own project I have a solution, so a big thanks for the discussion!

**Reply by jdx** [reply]
the docs certainly appear wrong: https://github.com/jdx/mise/blob/e291cc3802d027d9738b5060d9c68be8b20269e3/src/plugins/core/go.rs#L188-L202

I should have a new discussion on this topic with everything laid out soon though.

EDIT: I linked to the wrong code sample the first time

**Comment by brigitops** [comment]
> Looking at some of the history around this it seems there are reasons to set GOROOT explicitly even though `go` defaults it: [asdf-community/asdf-golang#28](https://github.com/asdf-community/asdf-golang/issues/28)
> 
> I'm not sure I follow why people want to set GOBIN and GOPATH to the asdf directory though

Agreed, I think it's more helpful to keep these consistent. If a complex project needed to isolate built binaries, I'd just use "go build -o" and put them somewhere else

**Reply by jdx** [reply]
so I was driving and I think it hit me why people asked for this with asdf but I don't see anyone in that thread spelling this out so it's a bit of a guess.

First, my understanding is there isn't any issue leaving GOPATH set to $HOME/go and switching go versions around. First I thought maybe the pkg directory might have version-specific stuff but that doesn't appear to be the case. Of course the src directory won't be version-specific either.

I think people want this because of the bin directory—which is why they also talk about GOBIN. Recall that asdf is shim-based where with mise using shims is only an option used by a minority (albeit significant minority) of users.

The behavior people are relying on is that when something is installed to `~/.asdf/installs/go/...`, it will create shims for those CLIs in `~/.asdf/shims`. **Those shims will get GOROOT set.** That is happening because GOPATH is set to a directory where asdf creates shims. If it went to HOME/go then no shims would be created and GOROOT would not be set.

I'll spell that out so it isn't clear: setting GOPATH is sort of a "trick" to get shims for go-installed CLIs which have GOROOT set. I don't think anyone actually needs or wants GOPATH to be modified except for setting GOROOT. I suspect this causes a lot of problems—in fact.

With mise, this isn't really a problem at least for `mise activate` or `mise exec` users. This is because `GOROOT` gets set for all commands, not just the shims—so effectively everything is shimmed with the correct GOROOT. Of course for people using shims this _would_ be a potential issue for users. Or if you use `mise activate` in your shell but your IDE relies on shims which use CLIs that need GOROOT to be set for go-installed CLIs.

I plan to write a larger doc on GOPATH/GOROOT but right now I'm really unsure what the right path forward is at least for GOPATH.

One option is to keep using GOPATH in ~/.local/share/mise/installs. This I feel is a pretty aggressive solution and probably causes a lot of problems by moving the src directory there. I don't like the current behavior. It does make shims that have GOROOT set though.

Option 2 is to switch to GOBIN instead. This will create new shims for go-installed CLIs as well but won't mess with the src directory. I prefer this to 1 but still dislike that every time you switch go versions you'll lose anything you go-installed. That said, perhaps the go [_backend_](https://mise.jdx.dev/dev-tools/backends/go.html) is more the right way to handle global CLIs anyways.

Option 3 would be to stop doing either. `mise activate` users will have good behavior, get to keep their CLIs under version changes, but people using shims may result in go-installed CLIs without GOROOT being set if they use shims.

I think I'm leaning towards 2 here. That said, I am not a go dev so will lean on the community for some guidance here.

Also this goes without saying but it's more evidence that my gut saying "just let `go env` figure out the default instead of explicitly setting GOROOT" was wrong. Evidently tools very much expect GOROOT to be set and are not deferring to `go env` or using compatible logic to determine GOROOT—if you ask me it sounds like those CLIs are the thing that is broken but given the popularity of that thread I suspect this is common even if it seems bad.

Also, if people have an example of a tool that relies on GOROOT and doesn't use `go env` to determine it, I would very much appreciate that since without that I fear I'm going out on a bit of a limb here but it seems to be the only thing that makes sense to me right now.

**Reply by brigitops** [reply]
I think that's all very fair. The only thing to add, really, is that go install itself breaks if the GOROOT doesn't match to go version of the go command. So even putting any broken third party tools aside, the correct behaviour can only be achieved IMHO if we either:

* always ignore the GOROOT set in the host environment, and isolate ourselves from it, so it behaves always as unset, because it's broken to explicitly set GOROOT other than the one matching the go tool chosen by mise anyway

or

* always override and set the GOROOT explicitly according to the go tool

Sounds like other people are running into a need for the second option. But for what broke my build, either option solves the thing.

I just can't think of a good reason to point GOROOT somewhere other than the go tool you're running, so why ever inherit it from the host environment? If somebody really wants to do that -- then sure we can given them a setting. But honestly, that's the weird workflow. It just... doesn't seem safe by default because there's no logic in sharing GOROOT across go installations. It's literally meant to point to the go installation.

Same with most tools, I would think? e.g. bad things happen when JAVE_HOME is pointing to something that is not, in fact, the java version you're using.

(edited for clarity)

**Reply by jdx** [reply]
> go install itself breaks

can you provide some color on this? an example would be helpful if you can think of one

> always ignore the GOROOT set in the host environment

"always" is a dangerous word. I think "by default" is probably the right way to go. What I'm leaning towards is not what happens currently where it is ignored if set but overwritten if set unless some _other_ setting like MISE_GO_SET_GOROOT is explicitly set to false or something.

I also don't know of a use-case where people would want a different GOROOT but I'm nervous about making assumptions like that.

**Comment by brigitops** [comment]
> > go install itself breaks
> 
> can you provide some color on this? an example would be helpful if you can think of one

Sure, so if the GoLand terminal GOROOT and path is pointing to go 1.21.6, and then I run mise where I've got it set for 1.22.0, the result is failure:

```
mise install
mise go@1.22.0 ✓ installed
mise exec -- go install gotest.tools/gotestsum@v1.11.0
# internal/goarch
compile: version "go1.21.6" does not match go tool version "go1.22.0"
```

What's happening, is that MISE is honouring the go tool version, but then it's also trying to honour the terminal GOROOT. But the two don't belong together, so the build fails.

This is reproducible after a mise implode.

I can sort-of fix this with:

```
[env]
MISE_GO_SET_GOROOT="true"
```

Which I didn't expect to make any difference, because documentation claims this is the default.

Though, the environment doesn't seem to apply first time, so I end up having to run my build twice. Second time, mise picks up that it should set the GOROOT. But that's another issue, I guess.

> > always ignore the GOROOT set in the host environment
> 
> "always" is a dangerous word. I think "by default" is probably the right way to go.

Fair, maybe I worded it poorly. All I really mean is:

*by default, mise should disregard the host GOROOT value, because it almost never makes sense for GOROOT to differ from the actual go tool being executed*

I'd go so far as to say it never makes sense at all. What use case for GOROOT to be misconfigured? But, it seems harmless to have an option for it.

> What I'm leaning towards is not what happens currently where it is ignored if set but overwritten if set unless some _other_ setting like MISE_GO_SET_GOROOT is explicitly set to false or something.

Not sure if we're on the same page about what the current behaviour is --  ignored if set? That would give me the behaviour I want!

I have reproduced the behaviour above with serveral mise implodes. I'm confident that the default behaviour is that it *does* inherit the GOROOT and that results in the weirdness above

**Reply by jdx** [reply]
ok I finished my new doc that should hopefully lay this out with a clear path forward: https://github.com/jdx/mise/discussions/1638

**Reply by brigitops** [reply]
Honestly, that's an awesome write-up!

Looks good to me. I'll read it again soon to make sure I didn't miss anything, and pick this up on #1638 itself (not a Rust developer but at the least, I can offer to test)

Thanks for the great discussion!

**Reply by jdx** [reply]
thank you! this was certainly a gap and while I'm surprised it's been out for this long I think this will give mise users a better DX for sure!

**Comment by Crocmagnon** [comment]
For what it's worth, since the Go added forward compatibility and toolchain management in go1.21 ([blog post](https://go.dev/blog/toolchain)) I stopped trying to mess my environment, just installed go1.21 once and called it a day.
Now I rely on setting the correct version in `go.mod` and I let go do its thing.

I still use mise for env variables and python though :)

**Reply by jdx** [reply]
yeah I saw that the other day. It certainly sounds like go devs are better off _not_ using mise (like my guidance for rust), but I suspect some will want to use it at least for possibly getting the first version of go installed.

Also, the go backend definitely still makes a lot of sense even with GOTOOLCHAIN.

I should dive into that more and come up with either good guidance in the docs or see if there is anything we can leverage to make mise integrate better with go.

This is a very good topic for a new discussion if anyone wants to start that off.

**Reply by Crocmagnon** [reply]
> getting the first version of go installed

I just want to side-note that installing go is surprisingly easy: all it takes is untaring an archive and adding a directory to the path 😀

**Reply by brigitops** [reply]
That's a great tip, and very fair. Only caveat and why I continue to investigate mise for the go toolchain, is that if you want serious reproducibility, you'd want it to pin the exact version, not just set a minimum. So that everyone is using the same version to build a given project, and that updating is a Merge Request on the project.

Unfortunately that only seems to happen if you force the toolchain with the environment variable approach, e.g.:

```
GOTOOLCHAIN=go1.20.4 go build
```

Unless I missed a feature of the toolchain in go.mod to do this, I'd still end up having to manage the environment to do this. Still, I might give it some more thought!

**Reply by jdx** [reply]
> I just want to side-note that installing go is surprisingly easy: all it takes is untaring an archive and adding a directory to the path 😀

I also think mise is overkill for just installing a single version of a tool, but there are certainly a lot of users that do like working that way. I think one reason is they like a single ".tool-versions" file that has everything in it so a vanilla system can just use `mise install` to be up and running.

It's worth thinking about that use-case, rust has a similar thing going on with rustup too. It may make sense to have the concept of "singleton tools" that are version-managers themselves that mise can integrate with more directly.

**Reply by Crocmagnon** [reply]
> pin the exact version
> [...]
> Unfortunately that only seems to happen if you force the toolchain with the environment variable approach

That’s a perfectly valid use case and as far as I can tell from the blog post and documentation, I believe you’re correct.

> I'd still end up having to manage the environment to do this

You could use mise to set the env variable 😛😁

Another side note: the Go team puts a lot of efforts into backwards compatibility as well, and there have been interesting developments since. In 1.22, the loop variable behavior changed ; but this is only applied when compiling with go1.22 on a project that lists go1.22 in the go.mod. When compiling a project listing go1.21 with a 1.22 toolchain, the old behavior is preserved.

As you said, it doesn’t solve the perfect reproducibility issue, but IMO it’s good enough for the majority of go devs.

> a single ".tool-versions" file that has everything in it so a vanilla system can just use `mise install` to be up and running

That's another valid use-case 👍🏻

**Comment by JamBalaya56562** [comment]
Both points of contention in this thread have since settled, and nobody said so here — the conversation moved to #1638 and stopped.

@brigitops' exact repro no longer reproduces. Measured on **v2026.8.3** with `go = "1.24"` and a foreign `GOROOT` inherited from the parent process, which is what the GoLand terminal does:

```console
$ GOROOT=/opt/sdk/go1.21.6 mise exec -- go env GOROOT
/home/jam/mise-go/data/installs/go/1.24.13

$ GOROOT=/opt/sdk/go1.21.6 mise exec -- go version
go version go1.24.13 linux/amd64
```

`GOROOT` and the `go` binary now agree, so `compile: version "go1.21.6" does not match go tool version "go1.22.0"` cannot happen from this cause.

The three variables are no longer treated alike. Measured the same way, reading `mise env`:

| parent environment | what mise sets |
| --- | --- |
| nothing set | `GOROOT` and `GOBIN` to the mise install path. **`GOPATH` is not set** |
| `GOROOT=/opt/sdk/go1.21.6` | `GOROOT` **overridden** with mise's |
| `GOBIN=/home/me/gobin` | **`GOBIN` not emitted** — the inherited one is left alone |
| `go.set_goroot = false` + `GOROOT` set | **`GOROOT` not emitted** |

So:

- **GOPATH** — no longer set at all. `go.set_gopath` is deprecated, defaults off, and warns if you set it (`src/plugins/core/go.rs:181`). That is @brigitops' "just leave GOPATH at `$HOME/go`" position, and it is the opposite of what @Crocmagnon originally asked for here — GOPATH is now deliberately not mise's business.
- **GOROOT** — set by default (`go.set_goroot`, default `true`) and it overrides an inherited value. The doc/implementation mismatch @brigitops ran into is gone: the documented default and the behaviour now agree, and `MISE_GO_SET_GOROOT="true"` is no longer needed as a workaround.
- **GOBIN** — deliberately asymmetric: set when explicitly enabled or when unset in the environment, otherwise the inherited value survives.

If you do want the IDE's `GOROOT` to win, the opt-out works:

```console
$ # [settings] go.set_goroot = false
$ GOROOT=/opt/sdk/go1.21.6 mise exec -- go env GOROOT
go: cannot find GOROOT directory: /opt/sdk/go1.21.6
```

(the path is bogus in this test, which is exactly why the error shows the inherited value is being deferred to).

One thing worth flagging for whoever picks this up: `go.set_gopath` carries a deprecation message reading *"Use `env._go.set_goroot` instead."*, which names `set_goroot` as the replacement for a `set_gopath` setting and points at an `env._go` directive that does not exist — the env directives are `file`, `module`, `path`, `source` and `venv`, and that string was the only occurrence of `env._go` in the repository. **It is never actually printed**, though: `warn_deprecated_now` only emits when `deprecated_warn_at` and `deprecated_remove_at` are also set, and this setting has neither. The warning you do see comes from `go.rs`'s `verify()` at install time and carries no guidance at all. Corrected text in #11799.
